File: src/hooks/nav2_hook.py
import logging
from typing import Any, Dict

import aiohttp

from hooks import Nav2Error
from providers.elevenlabs_tts_provider import ElevenLabsTTSProvider


async def start_nav2_hook(context: Dict[str, Any]):
    """
    Hook to start Nav2 process.

    Parameters
    ----------
    context : Dict[str, Any]
        Context dictionary containing configuration parameters.
    """
    base_url = context.get("base_url", "http://localhost:5000")
    map_name = context.get("map_name", "map")
    nav2_url = f"{base_url}/start/nav2"

    elevenlabs_provider: ElevenLabsTTSProvider = ElevenLabsTTSProvider()

    try:
        async with aiohttp.ClientSession() as session:
            async with session.post(
                nav2_url,
                json={"map_name": map_name},
                headers={"Content-Type": "application/json"},
                timeout=aiohttp.ClientTimeout(total=5),
            ) as response:

                if response.status == 200:
                    result = await response.json()
                    logging.info(
                        f"Nav2 started successfully: {result.get('message', 'Success')}"
                    )
                    elevenlabs_provider.add_pending_message(
                        "Navigation system has started successfully."
                    )
                    return {
                        "status": "success",
                        "message": "Nav2 process initiated",
                        "response": result,
                    }
                else:
                    try:
                        error_info = await response.json()
                    except Exception:
                        error_info = {"message": "Unknown error"}
                    error_msg = error_info.get("message", "Unknown error")
                    logging.error(
                        f"Failed to start Nav2: {error_msg} (status: {response.status})"
                    )
                    raise Nav2Error("Failed to start Nav2 navigation system")

    except aiohttp.ClientError as e:
        logging.error(f"Error calling Nav2 API: {e}")
        raise Nav2Error("Unable to communicate with Nav2 service") from e


async def stop_nav2_hook(context: Dict[str, Any]):
    """
    Hook to stop Nav2 process.

    Parameters
    ----------
    context : Dict[str, Any]
        Context dictionary containing configuration parameters.
    """
    base_url = context.get("base_url", "http://localhost:5000")
    nav2_url = f"{base_url}/stop/nav2"

    try:
        async with aiohttp.ClientSession() as session:
            async with session.post(
                nav2_url,
                headers={"Content-Type": "application/json"},
                timeout=aiohttp.ClientTimeout(total=5),
            ) as response:

                if response.status == 200:
                    result = await response.json()
                    logging.info(
                        f"Nav2 stopped successfully: {result.get('message', 'Success')}"
                    )
                    return {
                        "status": "success",
                        "message": "Nav2 process stopped",
                        "response": result,
                    }
                else:
                    try:
                        error_info = await response.json()
                    except Exception:
                        error_info = {"message": "Unknown error"}
                    error_msg = error_info.get("message", "Unknown error")
                    logging.error(
                        f"Failed to stop Nav2: {error_msg} (status: {response.status})"
                    )
                    raise Nav2Error("Failed to stop Nav2 navigation system")

    except aiohttp.ClientError as e:
        logging.error(f"Error calling Nav2 API: {e}")
        raise Nav2Error("Unable to communicate with Nav2 service") from e

File: internal/hooks/nav2_hook.go
package hooks

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"

	"go.uber.org/zap"

	"github.com/openmind/om1/internal/httpclient"
	"github.com/openmind/om1/internal/providers/tts"
)

const (
	nav2LocalBaseURL = "http://localhost:5000"
	nav2CloudBaseURL = "https://api.openmind.com/api/core/simulation/orchestrator"
	nav2DefaultMap   = "map"
)

func init() {
	RegisterHook("nav2_hook", "start_nav2_hook", (*Runner).startNav2Hook)
	RegisterHook("nav2_hook", "stop_nav2_hook", (*Runner).stopNav2Hook)
}

func (r *Runner) startNav2Hook(ctx context.Context, cfg, _ map[string]any) error {
	baseURL := nav2BaseURL(cfg)
	apiKey := stringVal(cfg, "api_key")
	mapName := stringValDefault(cfg, "map_name", nav2DefaultMap)
	nav2URL := baseURL + "/start/nav2"

	payload, err := json.Marshal(map[string]any{"map_name": mapName})
	if err != nil {
		return err
	}

	r.log.Info("nav2: starting", zap.String("url", nav2URL), zap.String("map_name", mapName))
	if err := nav2Post(ctx, nav2URL, apiKey, payload, 5*time.Second); err != nil {
		r.log.Error("nav2: failed to start", zap.Error(err))
		return fmt.Errorf("failed to start Nav2: %w", err)
	}

	r.log.Info("nav2: started successfully")
	r.nav2TTS(cfg).AddText("Navigation system has started successfully.")
	return nil
}

func (r *Runner) stopNav2Hook(ctx context.Context, cfg, _ map[string]any) error {
	baseURL := nav2BaseURL(cfg)
	apiKey := stringVal(cfg, "api_key")
	nav2URL := baseURL + "/stop/nav2"

	r.log.Info("nav2: stopping", zap.String("url", nav2URL))
	if err := nav2Post(ctx, nav2URL, apiKey, nil, 5*time.Second); err != nil {
		r.log.Error("nav2: failed to stop", zap.Error(err))
		return fmt.Errorf("failed to stop Nav2: %w", err)
	}

	r.log.Info("nav2: stopped successfully")
	return nil
}

func nav2BaseURL(cfg map[string]any) string {
	if base := stringVal(cfg, "base_url"); base != "" {
		return base
	}
	if useSim, _ := cfg["use_sim"].(bool); useSim {
		return nav2CloudBaseURL
	}
	return nav2LocalBaseURL
}

func nav2Post(ctx context.Context, url, apiKey string, body []byte, timeout time.Duration) error {
	reqCtx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()

	var reader io.Reader
	if body != nil {
		reader = bytes.NewReader(body)
	}
	req, err := http.NewRequestWithContext(reqCtx, http.MethodPost, url, reader)
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("x-api-key", apiKey)

	resp, err := httpclient.Default().Do(req)
	if err != nil {
		return fmt.Errorf("error calling Nav2 API: %w", err)
	}
	defer func() { _ = resp.Body.Close() }()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("%s", nav2ErrorMessage(resp))
	}
	return nil
}

func nav2ErrorMessage(resp *http.Response) string {
	var body struct {
		Message string `json:"message"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&body); err != nil || body.Message == "" {
		return "Unknown error"
	}
	return body.Message
}

func (r *Runner) nav2TTS(cfg map[string]any) *tts.ElevenLabsProvider {
	return tts.ElevenLabs(elevenLabsConfigFrom(cfg), r.log)
}
